packages/contracts/contracts/AaveAllocationStrategy.sol

```diff
@@ -0,0 +1,113 @@
+    function exchangeRateStored() public view returns (uint256) {
+        // Aave has a fixed exchange rate of 1:1 for aToken <-> underlying
+        // Interest is modeled by increasing balance of aTokens
+        // We calculate a virtual exchange rate based on the aToken balance and invested amount
+
+        if(totalInvested == 0) {
+            return 10**18;
+        }
+
+        // Safe Math not needed. aToken balance would need to be unfathomably high for the multiplication to overflow
+        return (aToken.balanceOf(address(this)) * 10**18) / totalInvested;
+    }
+
+    /// @dev Accrues interest. Not required for Aave protocol. Always returns true
+    function accrueInterest() external returns (bool) {
+        // Aaves interest accrual does not need to be called explicitly
+        // aToken.balanceOf() already contains the accrued interest
+        return true;
+    }
+
+    /// @dev Invest investAmount of underlying asset into Aave
+    function investUnderlying(uint256 investAmount) external onlyOwner returns (uint256) {
+        // Transfer underlying from caller to this contract
+        token.transferFrom(msg.sender, address(this), investAmount);
+        // Approve the Aave Lending Pool to access the underlying tokens
+        token.approve(address(lendingPoolCore), investAmount);
+        // Store the aToken balance of aTokens before deposit
+        uint256 aTotalBefore = aToken.balanceOf(address(this));
+        // Deposit investAmount of underlying asset into Aave
+        lendingPool.deposit(address(token), investAmount, 0);
+        // Calculate the difference in aToken balance after deposit
+        uint256 aTotalAfter = aToken.balanceOf(address(this));
+        // Check the aToken balance has increased after deposit
+        require (aTotalAfter >= aTotalBefore, "Aave minted negative amount!?");
+        // Update the total invested amount.
+        // No safe math needed since transfer would fail for invalid values
+        totalInvested += investAmount;
+        emit TotalInvested(totalInvested);
+        // Return the difference in aToken balance
+        return (investAmount * 10**18) / exchangeRateStored();
+    }
+
+    /// @dev Redeem redeemAmount from Aave
+    function redeemUnderlying(uint256 redeemAmount) external onlyOwner returns (uint256) {
+        // Store the aToken balance of aTokens before deposit
+        uint256 aTotalBefore = aToken.balanceOf(address(this));
+        // Redeem redeemAmount of underlying asset from Aave
+        aToken.redeem(redeemAmount);
+        // Calculate the difference in aToken balance after redeem
+        uint256 aTotalAfter = aToken.balanceOf(address(this));
+        // Check the aToken balance has decreased after redeem
+        require(aTotalAfter <= aTotalBefore, "Aave redeemed negative amount!?");
+        // Update totalInvested
+        totalInvested = (aTotalAfter * totalInvested) / aTotalBefore;
```

Comment:
This is great, it does the trick.

I really like that in the PR you highlighted the principle:

```
// We want to keep the exchange rate while updating the totalInvested. We calculate the newTotalInvested value we need to have the same exchange rate as before

oldExchangeRate = newExchangeRate
oldATokenBalance / oldTotalInvested = newATokenBalance / newTotalInvested      // solve for newTotalInvested
newATokenBalance  / (oldATokenBalance / oldTotalInvested) = newTotalInvested
newTotalInvested = (newATokenBalance * oldTotalInvested) / oldATokenBalance
```

Would you please take this principle to the comment too, that'd make the code self-explainatory.

Reply:
Added the comments to the contract source